src/neuron/container/soa_identifier.hpp

```diff
@@ -132,63 +137,62 @@ struct owning_identifier {
-            }
-            // We don't know how many people know the pointer `p`, so write a sentinel
-            // value to it and transfer ownership "elsewhere".
-            *p = invalid_row;
-            // This is to provide compatibility with NEURON's old nrn_notify_when_double_freed and
-            // nrn_notify_when_void_freed methods.
-            detail::notify_handle_dying(p);
-            // This is sort-of formalising a memory leak. In principle we could cleanup
-            // identifier_defer_delete_storage by scanning all our data structures and finding
-            // references to the pointers that it contains. In practice it seems unlikely that
-            // either this, or using std::shared_ptr just to avoid it, would be worth it.
-            if (detail::identifier_defer_delete_storage) {
-                detail::identifier_defer_delete_storage->emplace_back(p);
-            } else {
-                delete p;
-            }
+
+    void destroy() {
+        if (!m_ptr) {
+            // Nothing to be done.
+            return;
+        }
+
+        assert(m_data_ptr);
+        auto& data_container = *m_data_ptr;
+
+        // We should still be a valid reference at this point.
+        assert(m_ptr);
+        assert(m_ptr.current_row() < data_container.size());
+
+        // Prove that the bookkeeping works.
+        assert(data_container.at(m_ptr.current_row()) == m_ptr);
+
+        bool terminate = false;
+        // Delete the corresponding row from `data_container`
+        try {
+            data_container.erase(m_ptr.current_row());
+        } catch (std::exception const& e) {
+            // Cannot throw from unique_ptr release/reset/destructor, this
+            // is the best we can do. Most likely what has happened is
+            // something like:
+            //   auto const read_only_token = node_data.issue_frozen_token();
+            //   list_of_nodes.pop_back();
+            // which tries to delete a row from a container in read-only mode.
+            std::cerr << "neuron::container::owning_identifier<"
+                      << cxx_demangle(typeid(Storage).name())
+                      << "> destructor could not delete from the underlying storage: " << e.what()
+                      << " [" << cxx_demangle(typeid(e).name())
+                      << "]. This is not recoverable, aborting." << std::endl;
+            terminate = true;
         }
-        Storage* m_data_ptr{};
-    };
-    std::unique_ptr<std::size_t, deleter> m_ptr;
+        if (terminate) {
+            std::terminate();
+        }
+        // We don't know how many people know the pointer `p`, so write a sentinel
+        // value to it and transfer ownership "elsewhere".
+        m_ptr.set_current_row(invalid_row);
+
+        // This is to provide compatibility with NEURON's old nrn_notify_when_double_freed and
+        // nrn_notify_when_void_freed methods.
+        detail::notify_handle_dying(m_ptr);
```

Comment:
Sorry. I'm a bit shaky on this part. How does this relate to nrn_notify_when_double_freed? Is this a new notification (in the sense that there was not one before)?

Reply:
Nothing changed here; during this PR. What's happening is that the code from the deleter which is what got called when the `unique_ptr` was destroyed has been copied to the dtor of `owning_identifier`.

The semantics remains the same: when the `owning_identifier` is destroyed, the row in the `soa` is deallocated; and `detail::notify_handle_dying` is called. Which will call the following:
https://github.com/neuronsimulator/nrn/blob/40ae342d5231dd8412b4c7d4d2626711f25445f2/src/ivoc/ivoc.cpp#L114-L128